### DISCUSSION TITLE
Test Isolation with dotnetcore/dotnet5/dotnet6

### DISCUSSION THREAD

**Discussion by andez2000** [xunit/xunit#2322] [Question] [answered]
I did put a [question on StackOverflow](https://stackoverflow.com/questions/67849061/test-isolation-with-dotnetcore-dotnet5-dotnet6) the other day around the issue of isolating tests.

I wondered if it was possible - either through xunit or using some technique in `dotnet`.  Here I am focusing on `dotnetcore`, `dotnet5` and `dotnet6` - where the decision was to remove AppDomains.

But the question is basically, is it possible to have tests run under a clean context each time?

So something like the following deliberately coded class:

```
public class TheStaticFoo
{
    public TheStaticFoo() => Current = this;

    private static TheStaticFoo _current;

    public static TheStaticFoo Current
    {
        get => _current;

        private set
        {
            if (_current != null)
                throw new InvalidOperationException("Oh shucks, only one instance is allowed sorry!!!!");
            
            _current = value;
        }
    }
}
```

Called from a test:

```
public class UnitTest1
{
    private readonly TheStaticFoo foo;
    
    public UnitTest1() => foo = new TheStaticFoo();

    [Fact]
    public void Test1() => Assert.Equal(TheStaticFoo.Current, foo);
}
```

Which runs fine.

When a new method doing the same:

```
Fact]
public void AddingThisFails() => Assert.Equal(TheStaticFoo.Current, foo);
```

It is more the `System.Windows.Application` that has pushed me to write the above - something that I don't have control of - but want a clean environment where required in tests.

**Comment by bradwilson** [comment, marked as answer]
This isn't possible today. Generally the strategy I use personally is "don't use statics", which is a tough hill for some people to climb, especially because this probably changes the public API surface of your library. 😛

An in-between version would be that your internal implementation details don't use the statics but instead use an interface instance; in production, that interface instance would just wrap calls to the static, but in testing it would be a new clean instance per test.

What you're suggesting here for "clean context" is "restart the process". At the moment we have no plans to add this as a feature, because it would be extremely slow.

**Reply by andez2000** [reply]
Indeed - I hate static state.  Unfortunately, the static is [Application.Current](https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.current?view=net-5.0) under the `System.Windows` namespace, so one thread creates it then others cannot access properties on it.  I was running with a simple class with the static state that would cause issues in tests - to highlight the clean context.

If there are any potential workarounds I could check out I would be grateful if someone could point me at them.

**Reply by bradwilson** [reply]
In this case, you should probably use something like `[WinFormsFact]` from the [StaFact](https://github.com/AArnott/Xunit.StaFact) extensibility project. Any test decorated with this should have access to Windows Forms. (There are also a version for WPF, as well as generalized versions to either lock all code onto a single thread (`[UIFact]`) or onto the COM single threaded apartment (`[STAFact]`)).

**Reply by andez2000** [reply]
Hi Brad.  I am using that library.

Unfortunately, with dotnet5 at least, the Application never gets stripped down.  So was looking initially for a solution that could guarantee a clean runtime environment.  

I did initially ask about that here:
https://stackoverflow.com/questions/67647877/problems-running-multiple-xunit-tests-under-sta-thread-wpf

The problem is the Application class in the WPF environment causing me some headaches (whilst trying to be clever writing a navigation service).  I did write a hacky answer in response to my question.

**Reply by Luk164** [reply]
@bradwilson Hi, I have hit a similar problem and wanted to ask if something changed since the last time you answered here. I read something about version 3 being able to support separate processes for each test? I also see andez2000 found a bit of a hacky workaround so I may try that as well.

**Reply by bradwilson** [reply]
Nothing has changed. There are still no plans to reset the process for each test, though someone could (today, against v2) write such a runner if they wished. It'll be easier with v3, but it's probably doable today.

**Reply by Luk164** [reply]
@bradwilson Would you be willing to provide an example of such runner please? I have no idea where to start with something like that.

**Reply by andez2000** [reply]
> @bradwilson Hi, I have hit a similar problem and wanted to ask if something changed since the last time you answered here. I read something about version 3 being able to support separate processes for each test? I also see andez2000 found a bit of a hacky workaround so I may try that as well.

Yeah - my troubles were in WPF with the Application maintaining statics.  I would not recommend it though.

Maybe it would be cool to have the ability to optionally run certain test classes separately - in complete isolation.  Would be great if it was possible to write a custom test runner to do it if it is possible?

**Reply by bradwilson** [reply]
I'm sorry, I won't be able to provide a sample here. This would be an extremely involved process and I can't take my limited xUnit.net time away for writing such an in-depth sample. ☹️
